README.md

This repository contains an implementation of a branch-and-bound algorithm solving the integer *K_m,n* quadratic
optimization problem.

# Introduction

## Integer *K_m,n* quadratic programming problem

Let *A* is an *m*×*n* matrix of integer values.  
The goal is to compute the maximal value of *XAY*, where *X* and *Y* are vectors over {-1,1}.


## Branch-and-bound solution

The branch-and-bound solution of the problem is given in the II. section and the appendix of
[P. Diviánszky, E. Bene and T. Vértesi, "Qutrit witness from the Grothendieck constant of order four" In Physical Review A, 96(1):012113 (2017)](https://arxiv.org/abs/1707.04719)

The paper contains an application of the problem too.


## Implementation

The branch-and-bound solution of the problem is implemented in Haskell.

The program does the following steps:

1.  reads in the matrix *A* from a text file
2.  preprocessing: transform *A* to speed up the calculation without affecting the result
3.  generates an x86-64 assembly program which efficiently computes the maximum of *XAY*
4.  jumps to the generated assembly code
5.  the assembly code computes and prints the final result (and also partial results if requested)

    The full state of the assemby program is stored in registers,
    which drastically reduces the number of RAM accesses during the computation.
    This, and the use of SSE instructions speeds up the computation considerably.


## Limitations

-   The minimum of the matrix width and height should be less than 129.
-   The sum of the absolute values of the input matrix elements should be less than 2^31.
    The upper limit can be relaxed to 2^63 in some cases.

# Installation

Installation with [Cabal](https://www.haskell.org/cabal/):

    $ cabal install --installdir=$HOME

# Usage

## Basic usage

The most basic usage is to print the minimal value of *XAY*, where *A* is given in a text file:

    $ kmn-programming Wmat44.txt --silent
    1206540

## `--help`

The `--help` option prints a summary of the available options and commands:

    kmn-programming - specialized quadratic binary optimization

    Usage: kmn-programming (COMMAND | [-d|--delete] [-t|--transpose]
                           [-s|--sort SORTMETHOD] [-m|--multiply] [-p|--print]
                           [-s|--silent] [--levelin FILE] [--levelout FILE]
                           [-l|--level NAT] [--trace NAT] [-u|--unroll NAT]
                           [-a|--align NAT] [-o|--output FILE] FILE [--timeout ARG]
                           [--partial ARG])
      Maximalize sum of the input matrix multiplied by tensor products of two
      vectors of +-1 elements

    Available options:
      -h,--help                Show this help text
      -d,--delete              delete 0 rows and columns
      -t,--transpose           transpose matrix if it has more rows than columns
      -s,--sort SORTMETHOD     sort method - default is nosort
      -m,--multiply            multiply rows by +-1 to improve the first guess
      -p,--print               print matrix before optimization
      -s,--silent              print just the result
      --levelin FILE           precomputed levels input file
      --levelout FILE          levels output file
      -l,--level NAT           level - default is number of rows / 4
      --trace NAT              trace level - default is 0
      -u,--unroll NAT          unroll cycles - default is 4
      -a,--align NAT           code alignment - default is 8
      -o,--output FILE         output file
      --timeout ARG            timeout in seconds
      --partial ARG            do partial computation

    Available commands:
      sample                   print sample matrix
      test                     basic self-test
      timerandom               measure optimization time on random matrices

Note that for each command, there is a specific `--help` option too.

## Commands

There are 4 commands (modes of operation):

- maximize (this is the default, should not be written on the command line)
- `test`
- `sample`
- `timerandom`

The `test`, `sample` and `timerandom` modes are implemented for self-testing and benchmarks.

## Self-testing and benchmarking commands

The `test` command performs a basic self-test. The result should be `OK`:

    $ kmn-programming test
    OK

The `sample` command prints an n*(n-1) sized square matrix for each input n:

    $ kmn-programming sample 3
     2  3  3  0  3  3
     3  2  3  3  0 -3
     3  3  2 -3 -3  0
     0  3 -3  2  3 -3
     3  0 -3  3  2  3
     3 -3  0 -3  3  2

These matrices are described in Fishburn, P. C.; Reeds, J. A. (1994), Bell Inequalities, Grothendieck’s Constant, and Root Two, 7, SIAM Journal on Discrete Mathematics, pp. 48–56.
The maximal value of these matrices are known, so they are good for testing and benchmarking.

The produced matrix can be used in a separate step for benchmarking:

    $ kmn-programming sample 7 -o sample_7.txt
    $ time kmn-programming sample_7.txt 
    42 x 42
    levels: 10
    0000000000000000000001ffffffffff    924
    924

    real	0m42.987s
    user	0m42.980s
    sys	0m0.006s

The `timerandom` command can be used for benchmarking with random matrices.  
Type `--help` after `timerandom` to get some hints how to use this command:

    $ kmn-programming timerandom --help
    Usage: kmn-programming timerandom NAT NAT NAT NAT NAT NAT NAT NAT NAT NAT
      measure optimization time on random matrices

    Available options:
      NAT                      biggest integer in random matrices
      NAT                      smallest width
      NAT                      biggest width
      NAT                      width step
      NAT                      smallest height
      NAT                      biggest height
      NAT                      height step
      NAT                      smallest level
      NAT                      biggest level
      NAT                      repeat computation
      -h,--help                Show this help text

Example usage:

    $ kmn-programming timerandom 100 40 40 0 10 24 1 1 1 5
    (10,40,1)  2.302585092994046    -7.076399754920059
    (11,40,1)  2.3978952727983707    -7.10426931422738
    (12,40,1)  2.4849066497880004    -6.771162178173219
    (13,40,1)  2.5649493574615367    -6.889843064842471
    (14,40,1)  2.6390573296152584    -6.437137838157279
    (15,40,1)  2.70805020110221    -6.345387851709545
    (16,40,1)  2.772588722239781    -5.964019978283643
    (17,40,1)  2.833213344056216    -5.831543826032325
    (18,40,1)  2.8903717578961645    -5.635646650329268
    (19,40,1)  2.9444389791664403    -5.199033999407417
    (20,40,1)  2.995732273553991    -4.460960878254131
    (21,40,1)  3.044522437723423    -4.3522254540413625
    (22,40,1)  3.091042453358316    -3.5638282821343283
    (23,40,1)  3.1354942159291497    -3.0882569225539966
    (24,40,1)  3.1780538303479458    -2.822858809755754

Explanation of arguments:

100: the random matrix elements are chosen from {-100,-99,...,99,100} uniformly  
40: width of matrices in the first round  
40: width of matrices in the last round  
0: matrix with increment (no increment)  
10: height of matrices in the first round  
24: height of matrices in the last round  
1: matrix height increment between rounds  
1: level value in the first round (see later)  
1: level value in the last round  
5: make 5 samplings in each round

Explanation of the output:

There is one line of output for each round.

1st column: (matrix with, matrix height, level value)  
2nd column: logarithm of the matrix height  
3rd column: logarithm of the average runtime


## Maximization

Maximization is the default command. 

### Input file format

The input file should contain the elements of *A*, separated by whitespaces.  
Rows should be separated by one newline character.  
Elements of a row should be separated by spaces.  
Elements should be integers in decimal notation.

### Command line output formatting

Default output formatting (exmaple):

    $ kmn-programming Wmat44.txt 
    44 x 44
    levels: 11
    0000000000000000000007ffffffffff    1204990
    0000000000000000000007fffeffffff    1205496
    0000000000000000000007fffcfff7fc    1205584
    0000000000000000000007fffcdffffc    1205720
    0000000000000000000007fffaeffffb    1206022
    0000000000000000000006effcdffefc    1206372
    0000000000000000000006eff6fffffc    1206540
    1206540

44 x 44 is the matrix width and height.

`levels` is explained in the code generation section.

Then a line is printed each time when a better maximum is found.  
The second column is the new maximum.  
The first column is a hexadecimal number. The last 43 digit in the binary form of this number encodes the
last 43 elements of the *X* vector corresponding the maximum (1 --> +1, 0 --> -1).
The first element of *X* is always chosen to be 1.

At the end, the global maximum 1206540 is printed.

With the `--trace` option, one extra line is printed which part of the search space is evaluated.  
The search space is divided into 2^(*n*-1) equal parts where *n* is given by the user.  
The parts are numbered downwards, like 7. 6. 5. 4. 3. 2. 1. 0.:

    $ kmn-programming Wmat44.txt --trace 4
    44 x 44
    levels: 11
    7.
    0000000000000000000007ffffffffff    1204990
    0000000000000000000007fffeffffff    1205496
    0000000000000000000007fffcfff7fc    1205584
    0000000000000000000007fffcdffffc    1205720
    0000000000000000000007fffaeffffb    1206022
    6.
    0000000000000000000006effcdffefc    1206372
    0000000000000000000006eff6fffffc    1206540
    5.
    4.
    3.
    2.
    1.
    0.
    1206540

The `--silent` option suppresses all output but the found maximum:

    $ kmn-programming Wmat44.txt --silent
    1206540

There is another option affecting the output, `--print` which prints the preprocessed input matrix (see next section).


### Preprocessing

In the preprocessing phase the input matrix is transformed such that
the following steps run faster without affecting the found maximum.

The preprocessing phase has the following steps:

1.  Zero rows and columns are deleted if the `--delete` option is present.
2.  The matrix is transposed if the `--transpose` option is present and if
    the matrix has more rows than columns.
3.  The matrix rows are sorted if the `--sort` option is present.
    The following sorting methods are implemented:

    `asum`: sort the rows by decreasing absolute sum  
    `random`: the rows are premuted randomly, each time with a different seed  
    `random`NUM: the rows are premuted randomly with seed NUM  
    `b`: *not documented*
4.  Negate the rows of the matrix if the `--multiply` option is present.
    Not all rows are negated.
    The idea behind negation is that the first local maximum depends on the negation of the rows,
    but the global maximum does not.
    More subtrees can be pruned if we know that global maximum is at least *N*, if *N* is higher.
    `--multiply` tries to negate rows to increase the first local maximum.


### Code generation

The branch-and-bound algorithm skips subtrees in the searching phase if it cannot give a maximum
higher than the already found highest value.

Skipping subtrees has a cost, because the criteria for skipping should be evaluated.
We would like to reduce this cost, so on the first n level of the searching tree we do not
evaluate the skipping criteria.
The value of n can be given at the command line with the `--level` option.
If `--level` is missing, then n is one fourth of the number of rows.

The `--unroll` unroll option performs loop unrolling on the inner loop of generated assembly code.
Details:  
Each row is padded with 0 values such they will contain 2*ni* elements.
The inner row will be executed *i* times, evaluating 2*n* elements each time.
The number *n* can be given after `--unroll`.

The `--aling` option places certain labels in the assembly code on memory addresses
dividable by *n*. The number *n* can be given after `--align`.


### Distributed computation

The `--levelin`, `--levelout` and `--partial` options are implemented to support distributed computation
of the maximal value.

The `--timeout` option is also parsed but it is not implemented.

Here is an example, how to compute the maximum for a 90x90 sized matrix.

Assume that the matrix is given in the `mat90.txt` text file.

1.  Run

        $ kmn-programming mat90.txt --levelout levels.txt --level 0 --silent --partial 1/2^15

    2^15 is the number of subtasks. Any other power of 2 is OK.

2.  Run

        $ runhaskell SplitProblem.hs 15 mat90.txt --silent --levelin levels.txt >commands.sh

    After this, `commands.sh` will contain the subtasks.

4.  For each row in `commands.sh`, evaluate the row in a shell and collect the results.

5.  Compute the maximum of the results.




ChangeLog.md
# Revision history for kmn-programming

## 0.9.0.1  -- 2020-03-06

* Add Wmat44.txt
* Add SplitProblem.hs
* Add documentation in README.md

## 0.9.0.0  -- 2016-12-12

* First version.



KMNProgramming.hs
{-# language ViewPatterns #-}
{-# language RecursiveDo #-}
{-# language FlexibleInstances #-}
{-# language TypeSynonymInstances #-}
module KMNProgramming where

import Control.Monad hiding (unless)
import Control.Exception (evaluate)
import Foreign
import Foreign.Marshal.Unsafe

import CodeGen.X86

{- register usage
eax:    maximum value
rbx:    level pointer
rcx:    (aux) loop counter
rdx:    (aux) loop pointer
rsp:    (unused) stack pointer
rbp:    (aux) loop pointer
rsi:    (static) -16
rdi:    (static) vector pointer & start level
r8:     (static) trace level
r9:     (static) stop level
r10:    (static) debug flag
r11:    (static) level length
r12:    (static) loop length
r13:    result bits (high)
r14:    result bits (low)
r15:    (aux) vector norm
-}

foreign import ccall "static stdio.h &fflush" fflush :: FunPtr a

type T = Ptr Int32 -> Int32
foreign import ccall "dynamic" callPBII :: FunPtr T -> T
instance Callable T where dynCCall = callPBII

foreign import ccall "dynamic" callIO :: FunPtr (IO ()) -> IO ()
instance Callable (IO ()) where dynCCall = callIO

umes_ :: Bool -> Int -> Int -> Int -> Int -> Int -> [Int32] -> [[Int32]] -> (Int32, [Int32])
umes_ trr ali ali' tr uroll lev levs m = umes trr ali ali' tr uroll lev levs (length m) (length $ head m) m

{-# NOINLINE umes #-}
umes :: Bool -> Int -> Int -> Int -> Int -> Int -> [Int32] -> Int -> Int -> [[Int32]] -> (Int32, [Int32])
umes trr ali ali' tr uroll lev levs rows_ d_ = fun
  where
    fun (align' (4*uroll) -> a: as)
        = unsafeLocalState $ withArrayAligned (concat $ a: zipWith (++) ((: [0,0,0]) <$> (replicate (rows-length levs) 0 ++ levs)) as) (2^8) $ \r -> do
            x <- evaluate $ code r
            xs <- forM [d+d1*lev, d+d1*(lev+1)..d+d1*(rows-1)] $ peekElemOff r
            return (x, xs)

    rows = rows_ - 1
    d = d_ `modd` (4*uroll)
    d1 = d+4

    modd a b = b * ((a - 1) `div` b + 1)

    code :: Ptr Int32 -> Int32
    code = compile $ saveNonVolatile $ mdo
        mov rdx $ fromIntegral $ 4*d1*(rows-length levs)
        mloop <- label
        cmp rdx $ fromIntegral $ 4*d1*lev
        unless Z $ do
            save [rdi, rdx] $ do
                xor_ r10 r10
                call $ ipRelValue main
            mov (addr $ rdi + rdx - 16) eax
            sub rdx $ fromIntegral $ 4*d1
            jmp mloop

        xor_ rdx rdx
        mov r10 $ fromIntegral $ fromEnum trr
        call $ ipRelValue main
        jmp end

        main <- label
        lea r9  $ addr8 $ rdi + fromIntegral (4 * (rows + 1) * d1)
        lea r8  $ addr8 $ rdi + fromIntegral (4 * tr * d1)
        mov r11 $ fromIntegral $ 4*d1
        mov r12 $ fromIntegral $ d `div` (4*uroll)
        mov rsi (-16)
        xor_ rax rax
        xor_ r13 r13
        xor_ r14 r14
        lea rdi $ addr8 $ rdi + rdx
        mov rbx rdi
        jmp chk

        trace <- label
        save [rax, rcx, rdx, rbx, rbp, rsi, rdi, r8, r9, r10, r11, r12, r13, r14, r15] $ do
            leaData arg1 $ CString "%ld.\n"
            mov arg2 r14
            printfflush
        ret

        trace2 <- label
        save [rax, rcx, rdx, rbx, rbp, rsi, rdi, r8, r9, r10, r11, r12, r13, r14, r15] $ do
            leaData arg1 $ CString "%016lx%016lx    %d\n"
            mov arg2 r13
            mov arg3 r14
            mov (resizeOperand arg4) eax
            printfflush
        ret

        align ali
        chk <- label
        add rbx r11
        cmp rbx r9
        unless E $ mdo
            mov r15d $ addr $ rbx + rsi
            test r15d r15d
            if_ NZ (mdo
                mes
                cmp r15d eax
                j LE up) $ mdo
                cmp rbx r8
                unless NE $ call $ ipRelValue trace
            stc
            rcl r14 1
            when (rows > 64) $ rcl r13 1
            copyAdd
            jmp chk
        xor_ r15d r15d
        mes
        cmp r15d eax
        unless NA $ do
            mov eax r15d
            test r10 r10
            unless Z $ call $ ipRelValue trace2

        align ali
        up <- label
        sub rbx r11
        cmp rbx rdi
        unless E $ mdo
            test r14 1
            j Z zero
            xor_ r14 1
            copySub
            jmp chk
            zero <- label
            if rows > 64 then do
                shr r13 1
                rcr r14 1
              else do
                shr r14 1
            copyAdd
            jmp up
        ret

        end <- label
        return ()

    mes = mdo
        mov rcx r12
        mov rdx rdi
        pxor xmm2 xmm2
        align ali'
        doWhile NZ $ mdo
            replicateM_ uroll $ mdo
                movdqa xmm0 $ addr rdx
                sub rdx rsi
                movdqa xmm1 xmm0
                psrad  xmm0 31
                pxor  xmm1 xmm0
                psubd xmm2 xmm0
                paddd xmm2 xmm1
            sub rcx 1
        movdqa xmm1 xmm2
        psrldq xmm1 8
        paddd xmm2 xmm1
        movd ecx xmm2
        add r15d ecx
        psrldq xmm2 4
        movd ecx xmm2
        add r15d ecx

    copyAdd = copy $ mdo
        movdqa xmm0 $ addr rbp
        sub rbp rsi
        paddd  xmm0 $ addr rdx
        movdqa (addr rdx) xmm0
        sub rdx rsi

    copySub = copy $ mdo
        movdqa xmm0 $ addr rbp
        sub rbp rsi
        movdqa xmm1 $ addr rdx
        paddd xmm0 xmm0
        psubd xmm1 xmm0
        movdqa (addr rdx) xmm1
        sub rdx rsi

    copy body = mdo
        mov rcx r12
        mov rbp rbx
        mov rdx rdi
        align ali'
        doWhile NZ $ mdo
            replicateM_ uroll body
            sub rcx 1

    printfflush = do
        xor_ rax rax
        callFun rbp printf
        xor_ arg1 arg1
        callFun rbp fflush

align' n xs = (++ pad) <$> xs
  where
    pad = replicate ((-length (head xs)) `mod` n) 0

withArrayAligned bs al cont = allocaBytesAligned (sizeOf (head bs) * length bs) al $ \r -> do
    forM_ (zip [0..] bs) $ uncurry $ pokeElemOff r
    cont r

save rs c = do
    sequence_ [push r | r <- rs]
    c
    sequence_ [pop r | r <- reverse rs]



Main.hs
{-# language PatternGuards #-}
{-# language ViewPatterns #-}
{-# language NoMonomorphismRestriction #-}
import Data.Function
import Data.List
import Data.Monoid
import Data.Maybe
import Data.Time.Clock
import Control.Arrow
import Control.Monad
import Control.Concurrent
import Options.Applicative
import System.IO
import System.Random
import System.Random.Shuffle

import KMNProgramming

data Method = Masum | Mrandom (Maybe Int) | Mb | Mnosort
 deriving (Read, Show, Eq, Ord)

readMethod = eitherReader $ \s -> case s of
    "nosort" -> Right Mnosort
    "asum" -> Right Masum
    "b" -> Right Mb
    (splitAt 6 -> ("random", s))
        | "" <- s -> Right $ Mrandom Nothing
        | [(i, "")] <- reads s -> Right $ Mrandom $ Just i
    _ -> Left $ "available sort methods: asum random randomSEED b nosort"

main :: IO ()
main = do
    hSetBuffering stdout LineBuffering
    join $ execParser opts
  where
    opts = info (helper <*> options)
      ( fullDesc
     <> progDesc "Maximalize sum of the input matrix multiplied by tensor products of two vectors of +-1 elements"
     <> header "kmn-programming - specialized quadratic binary optimization")

    options :: Parser (IO ())
    options = 
        hsubparser
           ( command "sample" sampleCommand
          <> command "test" testCommand
          <> command "timerandom" randomCommand
           )
     <|> computeOptions

    sampleCommand = flip info (progDesc "print sample matrix") $ printTestMat
        <$> optional (strOption $ short 'o' <> long "output" <> metavar "FILE" <> help "output file" <> action "filenames")
        <*> argument auto (metavar "NAT")

    testCommand = flip info (progDesc "basic self-test") $ testTestMat
        <$> optional (argument auto $ metavar "NAT")

    randomCommand = flip info (progDesc "measure optimization time on random matrices") $ timeRandom
--        <$> optional (strOption $ short 'o' <> long "output" <> metavar "FILE" <> help "output file" <> action "filenames")
        <$> argument auto (metavar "NAT" <> help "biggest integer in random matrices")
        <*> argument auto (metavar "NAT" <> help "smallest width")
        <*> argument auto (metavar "NAT" <> help "biggest width")
        <*> argument auto (metavar "NAT" <> help "width step")
        <*> argument auto (metavar "NAT" <> help "smallest height")
        <*> argument auto (metavar "NAT" <> help "biggest height")
        <*> argument auto (metavar "NAT" <> help "height step")
        <*> argument auto (metavar "NAT" <> help "smallest level")
        <*> argument auto (metavar "NAT" <> help "biggest level")
        <*> argument auto (metavar "NAT" <> help "repeat computation")

    computeOptions = compute
        <$> switch (short 'd' <> long "delete" <> help "delete 0 rows and columns")
        <*> switch (short 't' <> long "transpose" <> help "transpose matrix if it has more rows than columns")
        <*> optional (option readMethod $ short 's' <> long "sort" <> metavar "SORTMETHOD" <> help "sort method - default is nosort" <> completeWith ["asum","random","b","nosort"])
        <*> switch (short 'm' <> long "multiply" <> help "multiply rows by +-1 to improve the first guess")
        <*> switch (short 'p' <> long "print" <> help "print matrix before optimization")
        <*> switch (short 's' <> long "silent" <> help "print just the result")
        <*> optional (strOption $ long "levelin" <> metavar "FILE" <> help "precomputed levels input file" <> action "filenames")
        <*> optional (strOption $ long "levelout" <> metavar "FILE" <> help "levels output file" <> action "filenames")
        <*> optional (option auto $ short 'l' <> long "level"  <> metavar "NAT" <> help "level - default is number of rows / 4")
        <*> (fromMaybe 0 <$> optional (option auto $ long "trace"  <> metavar "NAT" <> help "trace level - default is 0" <> completeWith ["0"]))
        <*> (fromMaybe 4 <$> optional (option auto $ short 'u' <> long "unroll" <> metavar "NAT" <> help "unroll cycles - default is 4" <> completeWith ["4"]))
        <*> (fromMaybe 8 <$> optional (option auto $ short 'a' <> long "align" <> metavar "NAT" <> help "code alignment - default is 8" <> completeWith ["8"]))
        <*> optional (strOption $ short 'o' <> long "output" <> metavar "FILE" <> help "output file" <> action "filenames")
        <*> (   {-Right . mkMatrix . filter (not . null) . concatMap splitSC <$ argument (readChars "matrix") (metavar "MATRIX") <*> many (argument (maybeReader readMat) (metavar "INT"))
            <|> -}Left  <$> argument str (metavar "FILE" <> action "filenames")
            )
        <*> optional (option auto $ long "timeout" <> help "timeout in seconds")
        <*> optional (option auto $ long "partial" <> help "do partial computation")

readChars c = maybeReader f where
    f c' | c == c' = Just ()
    f _ = Nothing

readMat s@(_:_) = Just s
readMat _ = Nothing

mkMatrix s = case span (/= ",") s of
    (i, ",": s) -> i: mkMatrix s
    (i, []) -> [i]

splitSC s = case span (/= ',') s of
    (i, ',': s) -> i: ",": splitSC s
    (i, []) -> [i]

checkPred err p | p = return ()
checkPred err _ = error err

data Partial = Partial Int Int
    deriving (Eq, Ord, Show)

instance Read Partial where
    readsPrec _ s
        | [(i,'/':'2':'^':s)] <- reads s
        , [(j,s)] <- reads s
        = [(Partial i j, s)]
    readsPrec _ _ = []

compute :: Bool -> Bool -> Maybe Method -> Bool -> Bool -> Bool -> Maybe FilePath -> Maybe FilePath -> Maybe Int -> Int -> Int -> Int -> Maybe FilePath -> Either FilePath [[String]] -> Maybe Int -> Maybe Partial -> IO ()
compute del transp met mult printmat silent levelin levelout level_ tra uroll ali out fname timeout partial = do
    gen <- newStdGen
    s <- either (fmap (filter (not . null) . map words . lines) . readFile) return fname

    levs <- case levelin of
        Nothing -> return []
        Just f -> map read . lines <$> readFile f

    let vs = map (map read) s :: [[Integer]]

        mat_ = (if transp then transpose' else id)
            . (if del then transpose . filter (any (/=0)) . transpose . filter (any (/=0)) else id)
            $ map fromIntegral <$> vs

        transpose' xs = if length xs <= length t then xs else t
          where
            t = transpose xs

        ns@ ~[width] = nub $ length <$> mat_
        output = maybe putStrLn writeFile out

        method = case met of
            Nothing -> id
            Just Mnosort -> id
            Just Masum  -> sortByASum
            Just Mb     -> sortByASum'
            Just (Mrandom mi) -> \xs -> shuffle' xs (length xs) $ maybe gen mkStdGen mi

        mat = (if mult then multguess else id)
            . method
            $ mat_

        mat' = maybe mat (\(Partial i j) -> (mkPartial (2^j-i) j mat)) partial

        nx = length mat'
        level = fromMaybe (min (nx - length levs - 1) $ round $ fromIntegral nx / 4) level_

        (res, levs') = umes_ (not silent) ali ali tra uroll level levs mat'

    checkPred "matix row lengths differ" $ length ns == 1
    checkPred "number of rows should be less than 129" $ nx <= 128
    checkPred "alignment should be 1,2,4,8,16,..." $ ali `elem` map (2^) [0..10]
    checkPred "unroll should be greater than 1" $ uroll >= 1
--    checkPred ("unroll " ++ show uroll ++ " less or equal then the width " ++ show width $ uroll <= width
    checkPred "level should not be negative" $ level >= 0
    checkPred "level should be less than the number of rows" $ level < nx
    checkPred "trace level should not be negative" $ tra >= 0
    checkPred "trace level should be less or equal than level" $ tra <= level

    when (not silent) $ do
        when (asum (concat vs) >= 2^31) $ output "!!! warning: overflow may happen !!!"
        output $ show nx ++ " x " ++ show width
        output $ "levels: " ++ show level
        when printmat $ output $ showMat mat'
    tid <- myThreadId
    case timeout of
        Nothing -> return ()
        Just t -> void $ forkIO $ do
            threadDelay $ 1000000 * t
            killThread tid
    case levelout of
        Nothing -> return ()
        Just f -> writeFile f $ unlines $ map show levs'
    output $ show res

mkPartial i 0 mat = mat
mkPartial i j (x: y: mat) = mkPartial (i`div`2) (j-1) (zipWith (if odd i then (+) else (-)) x y: mat)

timeRandom :: Int -> Int -> Int -> Int -> Int -> Int -> Int -> Int -> Int -> Int -> IO ()
timeRandom r y y' sy x x' sx (fromIntegral -> z) (fromIntegral -> z') rep = forM_ (f $ zip [x, x+sx..x'] [y, y+sy..y']) $ \((i, j), z) -> do
    putStr $ show (i,j,z) ++ "  " ++ show (log $ fromIntegral i) ++ "    "
    mats <- replicateM rep $ replicateM i $ replicateM j $ fromIntegral <$> randomRIO (-r, r)
    let opt = umes False 8 8 0 1 z [] i j
    t1 <- getCurrentTime
    forM_ mats $ \mat -> return $! fst $ opt mat
    t2 <- getCurrentTime
    putStrLn $ show $ log $ (realToFrac (diffUTCTime t2 t1) :: Double) / fromIntegral rep
  where
    f xs = zip xs $ map round [z, z + (z' - z)/(fromIntegral (length xs) - 1) ..]

printTestMat :: Maybe FilePath -> Int -> IO ()
printTestMat out n = maybe putStrLn writeFile out $ showMat $ testMat n

testTestMat :: Maybe Int -> IO ()
testTestMat n_
    | f n == res = putStrLn "OK"
    | otherwise = error "fatal error"
  where
    n = fromMaybe 6 n_
    m = testMat n
    res = fst $ umes False 8 8 0 1 (length m `div` 4) [] (length m) (length $ head m) m
    f 2 = 4
    f 3 = 36
    f 4 = 120
    f 5 = 280
    f 6 = 540
    f 7 = 924

testMat n = [[(if i == j then 1 else 3) * inner i j | i<-fs] | j<-fs]
  where
    fs = [ins b   1  $ ins a 1 $ replicate (n-2) 0 | a <- [0..n-2], b <- [a+1..n-1]]
      ++ [ins b (-1) $ ins a 1 $ replicate (n-2) 0 | a <- [0..n-2], b <- [a+1..n-1]]

    ins i a (splitAt i -> (as, bs)) = as ++ a: bs

showMat = unlines . map unwords . transpose . map (pad . map (g . show)) . transpose
  where
    pad xs = map f xs
      where
        f = take n . (++ repeat ' ')
        n = maximum $ 0: map length xs
    g s@('-': _) = s
    g s = ' ': s

sortByASum = map snd . sortBy (flip compare `on` fst) . map (asum &&& id)

asum = sum . map abs

inner a b = sum $ zipWith (*) a b

sortByASum' [] = []
sortByASum' xs = reverse $ a: f a as
  where
    (a, as) = snd $ maximumBy (flip compare `on` fst) $ map (asum . fst &&& id) $ getOut xs

    f _ [] = []
    f x xs = a: f a as
      where
        (a, as) = snd $ maximumBy (flip compare `on` fst) $ map (abs . inner x . fst &&& id) $ getOut xs

    getOut :: [a] -> [(a, [a])]
    getOut xs = [(x, as ++ bs) | (as, x:bs) <- zip (inits xs) (tails xs)]

multguess (x: xs) = x: f x xs
  where
    f _ [] = []
    f v (x: xs)
        | asum a >= asum b = x: f a xs
        | otherwise = map negate x: f b xs
      where
        a = zipWith (+) v x
        b = zipWith (-) v x

